index.ts

```diff
@@ -253,34 +255,70 @@ export default class Analytics {
       message.messageId = `node-${md5(JSON.stringify(message))}-${uuid()}`;
     }
 
-    this.queue.push({ message, callback });
+    this.messageQueue.push({ message, callback });
 
     if (!this.flushed) {
       this.flushed = true;
       this.flush();
       return;
     }
 
-    const hasReachedFlushAt = this.queue.length >= this.flushAt;
-    if (hasReachedFlushAt) {
-      this.logger.debug('flushAt reached, trying flush...');
+    const isDivisibleByFlushAt = this.messageQueue.length % this.flushAt === 0;
+    if (isDivisibleByFlushAt) {
+      this.logger.debug(
+        `flushAt reached, messageQueueLength is ${this.messageQueue.length}, trying flush...`
+      );
       this.flush();
-    }
-
-    if (this.flushInterval && !this.timer) {
+    } else if (this.flushInterval && !this.timer) {
+      // only start a timer if there are dangling items in the message queue
       this.logger.debug('no existing flush timer, creating new one');
       this.timer = setTimeout(this.flush.bind(this), this.flushInterval);
     }
   }
 
   /**
-   * Flushes the message queue to the server immediately if a flush is not already in progress.
+   * Requests a flush to be executed ASAP -  if a flush is not already in progress enqueues a new flush for when the current flush is finished.
    */
   async flush(callback: AnalyticsFlushCallback = () => {}): Promise<FlushResponse> {
+    let boundFlush;
+    const promisedFlush = new Promise<FlushResponse>((resolve) => {
+      boundFlush = async () => {
+        const flushResponse = await this.executeFlush(callback);
+        resolve(flushResponse);
+      };
+    });
+    this.flushQueue.push(boundFlush);
```

Comment:
The enqueued flush could be waiting for a long time (until a new event is enqueued) if there's an inflight flush sending the last entries in the queue.

Reply:
yup. the reason it matters that we wait for that specific flush (and not an earlier one) is each flush can contain a unique callback. We only want to return resolved when that callback has been fired, otherwise we could unintentionally break a consumer's app flow.

Reply:
Yes but there's a case where an inflight flush completes and then doesn't process the flush queue, leaving this enqueued flush to sit indefinitely, not just until earlier flushes have finished.